src/acl.c

```diff
@@ -1506,6 +1506,8 @@ user *ACLGetUserByName(const char *name, size_t namelen) {
  * If the selector can access the key, ACL_OK is returned, otherwise
  * ACL_DENIED_KEY is returned. */
 static int ACLSelectorCheckKey(aclSelector *selector, const char *key, int keylen, int keyspec_flags) {
+    int is_sort_pattern = keyspec_flags & CMD_KEY_SORT_PATTERN;
```

Comment:
I think it would be a smaller change to just have a small ACL check like here https://github.com/redis/redis/blob/unstable/src/sort.c#L246. It doesn't flow our normal ACL machine, but it is an error we can check and throw early. We can also throw a more straight forward error like, "The GET/BY option requires all key permissions to execute".

Reply:
@madolson I have several issues with your suggestion:
1. Although we all agreed this is not important - this will again not work with dryrun which I think is sub-optimal
2. the change will require (like option 2) to propagate the selector to the sort command  (since we might have validated the sort with a selector that dos NOT have full key access) or move the entire sort command authorization inside the sort, which is getting closer to option 3
3. I think mixing ACL logic with application logic is better be prevented. I like the fact that the sort is kept unaware of the ACL which is the main advantage of this option (unlike Options 2+3)

Reply:
The only thing I'm thinking about is the error will look like:
```
The user does not have access to the 'foo_*->whatever' key
```
but I suppose we can just document that.

Reply:
@madolson I can try using '-1' for the key index and make a generic error msg in that case

Reply:
I just think we have to accept it. It breaks a bunch of contracts including the module contract.

Reply:
@oranagra I understand all that is said. 
1. IMHO the DRYRUN command is important not only for debug purpose. given the potential complexity ACL can now get when we introduce selectors, I found myself using the DRYRUN command often in order to prevent making unnecessary changes to the data. also it never provides good customer experience having API which provides false positive.
2. checking if the command has unrestricted key access inside the sort will require evaluating the command and arguments again in the sort, which is exactly what this solution was trying to avoid (mixing application logic with the ACL logic) I think that the sort performance would probably not be that effected by it, but future commands which might except access patterns would also need to have a specialized handling of ACL matching.

I do not have a strong argument against what you suggest, I am just pointing out what I think are the advantages in this proposal.

Reply:
regarding 1, I accept the argument that DRYRUN is a feature, not just for testing, but i still argue that it's ok that it can't cover some cases. for instance, it also can't cover EVAL scripts and possibly some module commands.
i argue that this SORT feature is esoteric, which is why we decided not to properly support it, the reason being that we don't wanna pay the code complexity price, so the current change in this PR seems to beat that purpose.

regarding 2, we are evaluating the command argument in SORT anyway, and even if we have to evaluate them one more time, that's won't be an impact on performance.
i'm aiming to have most of the logic about this hack in one place (even if it'll be a lot of code, the complexity is low since it's not spread all over redis).

so adding one short and generic function in acl.c with clear inputs and output, and no SORT specific logic, and a big pile of code in sortCommand, seems like the right solution to me.

and again, i'm willing to overlook the fact that `#` doesn't require key access if it makes a big impact on the complexity of the code.

Reply:
> regarding 1, I accept the argument that DRYRUN is a feature, not just for testing, but i still argue that it's ok that it can't cover some cases. for instance, it also can't cover EVAL scripts and possibly some module commands.

I think that the DRYRUN should not validate commands which it cannot validate - it should return error or some clear notification that this (EVAL, EVALSHA etc...) cannot be verified. for the Sort case, I think that the DRYRUN can also use the same logic done in sort in order to validate the command. This was easier to achieve in this implementation, but we can also add this test in DRYRUN for sort. HOWEVER if we want to ignore this for sort we can also just refuse to validate the sort like other "unvalidatable" commands.

I agree with all the of the arguments. this can be done inside the sort with no major performance degradation.